### PR TITLE
[Email] Handle Gmail Exception

### LINKED ISSUE #412: Exception handling for GmailAPI in EmailSkill
**Project**
Virtual Assistant, Skills

**Description**
Need to add proper exception handling for GmailAPI calls in EmailSkill

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
## Description
Add support for gmail exception handling.

## Related Issue
Exception handling for GmailAPI in EmailSkill 
fixes #412

## Testing Steps
1. Remove gmail scope in google account.
2. Try show emails with google account
3. Bot reply: Sorry, there is something wrong with the bot at the moment. The developer has been notified about it so you'll be able to perform the action again very soon.

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the appropriate unit tests
- [ ] I have updated related documentation

If this contains changes that needs to be replicated between the Enterprise Template <-> Virtual Assistant
- [ ] A duplicate issue is filed to track future  work.

If you have updated responses or `.lu` files:
- [ ] All languages have been updated
- [ ] You have tested deployment with your new models

### CODE REVIEW

**Review by lauren-mills — APPROVED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 1]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
         /// <inheritdoc/>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs]: remove this #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 2]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
...
-            }
-
-            // set the reply subject
-            forward.Subject = "Fwd: " + originMessage.Subject;
+                // getOriginalMessage
+                var (originMessage, threadId) = await this.GetMessageById(id);
+                var forward = new MimeMessage();
+                foreach (var recipient in recipients)
+                {
+                    forward.To.Add(new MailboxAddress(recipient.EmailAddress.Address));
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs:87]: is 'forward.To' going to be initialized once 'forward' is initialized? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 3]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
...
+                forward.Subject = EmailCommonStrings.Forward + originMessage.Subject;
 
-            if (!string.IsNullOrEmpty(originMessage.MessageId))
-            {
-                forward.References.Add(originMessage.MessageId);
-            }
+                // construct the References headers
+                foreach (var mid in originMessage.References)
+                {
+                    forward.References.Add(mid);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs:96]: same comment for forward.References. do you not have to instantiate it before using it? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 4]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
...
+                    forward.To.Add(new MailboxAddress(recipient.EmailAddress.Address));
+                }
 
-            // construct the References headers
-            foreach (var mid in originMessage.References)
-            {
-                forward.References.Add(mid);
-            }
+                // set the reply subject
+                forward.Subject = EmailCommonStrings.Forward + originMessage.Subject;
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs]: This is assuming the subject is always going to be 
Forward: {Subject}

I think it's better define the whole string as a template string in resource
Forward: {0}

and in here use 

string.format(EmailCommonStrings.ForwardSubjectTemplate, originMessage.Subject) #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 5]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
...
-            foreach (var recipient in recipients)
+            try
             {
-                forward.To.Add(new MailboxAddress(recipient.EmailAddress.Address));
-            }
-
-            // set the reply subject
-            forward.Subject = "Fwd: " + originMessage.Subject;
+                // getOriginalMessage
+                var (originMessage, threadId) = await this.GetMessageById(id);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs]: originalMessage is a little better than originMessage #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs [hunk 6]
```diff
@@ -76,265 +78,289 @@ public static GmailService GetServiceClient(GoogleClient config, string token)
...
             {
-                Raw = Base64UrlEncode(forward.ToString() + content),
-                ThreadId = threadId,
-            }, "me").ExecuteAsync();
-        }
+                throw GoogleClient.HandleGoogleAPIException(ex);
+            }
+}
 
         /// <inheritdoc/>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs]: remove this #Resolved

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs (hunk 2 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs:87]: Yes, I checked the source code. #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs (hunk 3 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/GoogleAPI/GMailService.cs:96]: Checked the source code, the To and References are initialized in constructor and managed inside the class. And they only have get function, so we have no way to initialize them. #Resolved

**Review by lzc850612 — APPROVED** [review]

**Pull request merged**